# Customer Churn & Retention Analysis

## 1. Data Loading

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/telco.csv")

In [3]:
df.head()

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,...,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,...,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,...,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,...,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,...,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges


## 2. Data Understanding & Validation

In [4]:
df.shape

(7043, 50)

In [5]:
df.columns.tolist()

['Customer ID',
 'Gender',
 'Age',
 'Under 30',
 'Senior Citizen',
 'Married',
 'Dependents',
 'Number of Dependents',
 'Country',
 'State',
 'City',
 'Zip Code',
 'Latitude',
 'Longitude',
 'Population',
 'Quarter',
 'Referred a Friend',
 'Number of Referrals',
 'Tenure in Months',
 'Offer',
 'Phone Service',
 'Avg Monthly Long Distance Charges',
 'Multiple Lines',
 'Internet Service',
 'Internet Type',
 'Avg Monthly GB Download',
 'Online Security',
 'Online Backup',
 'Device Protection Plan',
 'Premium Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Streaming Music',
 'Unlimited Data',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charge',
 'Total Charges',
 'Total Refunds',
 'Total Extra Data Charges',
 'Total Long Distance Charges',
 'Total Revenue',
 'Satisfaction Score',
 'Customer Status',
 'Churn Label',
 'Churn Score',
 'CLTV',
 'Churn Category',
 'Churn Reason']

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 50 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   object 
 1   Gender                             7043 non-null   object 
 2   Age                                7043 non-null   int64  
 3   Under 30                           7043 non-null   object 
 4   Senior Citizen                     7043 non-null   object 
 5   Married                            7043 non-null   object 
 6   Dependents                         7043 non-null   object 
 7   Number of Dependents               7043 non-null   int64  
 8   Country                            7043 non-null   object 
 9   State                              7043 non-null   object 
 10  City                               7043 non-null   object 
 11  Zip Code                           7043 non-null   int64

In [7]:
df.isnull().sum()

Customer ID                             0
Gender                                  0
Age                                     0
Under 30                                0
Senior Citizen                          0
Married                                 0
Dependents                              0
Number of Dependents                    0
Country                                 0
State                                   0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Population                              0
Quarter                                 0
Referred a Friend                       0
Number of Referrals                     0
Tenure in Months                        0
Offer                                3877
Phone Service                           0
Avg Monthly Long Distance Charges       0
Multiple Lines                          0
Internet Service                  

In [8]:
df["Customer ID"].duplicated().sum()

np.int64(0)

## 3. Overall Churn Analysis

### 3.1 Customer Status

In [9]:
df["Customer Status"].value_counts()

Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64

### 3.2 Churn label

In [10]:
df["Churn Label"].value_counts()

Churn Label
No     5174
Yes    1869
Name: count, dtype: int64

### 3.3 Overall Churn Rate

In [11]:
churn_rate = (df["Churn Label"] == "Yes").mean() * 100
churn_rate

np.float64(26.536987079369588)

## 4. Customer Profile Analysis

### 4.1 Senior Citizen

In [12]:
df.groupby("Senior Citizen")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

Senior Citizen
No     23.606168
Yes    41.681261
Name: Churn Label, dtype: float64

### 4.2 Married

In [14]:
df.groupby("Married")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

Married
No     32.957979
Yes    19.664903
Name: Churn Label, dtype: float64

### 4.3 Dependents

In [15]:
df.groupby("Dependents")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

Dependents
No     32.551699
Yes     6.515058
Name: Churn Label, dtype: float64

### 4.4 Gender

In [16]:
df.groupby("Gender")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

Gender
Female    26.920872
Male      26.160338
Name: Churn Label, dtype: float64

## 5. Contract & Tenure Analysis

### 5.1 Churn by Contract

In [17]:
df.groupby("Contract")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100     
)

Contract
Month-to-Month    45.844875
One Year          10.709677
Two Year           2.549124
Name: Churn Label, dtype: float64

### 5.2 Churn by Tenure Group

In [18]:
df["Tenure Group"] = pd.cut(
    df["Tenure in Months"],
    bins=[-1, 12, 24, 36, 48, 60, float("inf")],
    labels=["0-12", "13-24", "25-36", "37-48", "49-60", "61+"]
)

In [19]:
df.groupby("Tenure Group", observed=True)["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

Tenure Group
0-12     47.438243
13-24    28.710938
25-36    21.634615
37-48    19.028871
49-60    14.423077
61+       6.609808
Name: Churn Label, dtype: float64

## 6. Service & Subscription Analysis

### 6.1 Internet Service

In [ ]:
df.groupby("Internet Service")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 6.2 Internet Type

In [ ]:
df.groupby(
    df["Internet Type"].fillna("No Internet")
)["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100 
)

### 6.3 Online Security

In [ ]:
df.groupby("Online Security")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 6.4 Premium Tech Support

In [ ]:
df.groupby("Premium Tech Support")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 6.5 Unlimited Data

In [ ]:
df.groupby("Unlimited Data")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 6.6 Streaming Services

In [ ]:
streaming_cole = [
    "Streaming TV",
    "Streaming Movies",
    "Streaming Music"
]

for col in streaming_cole:
    print(f"\n{col}")
    print(
        df.groupby(col)["Churn Label"].apply(
            lambda x: (x == "Yes").mean() * 100     
        )
    )

## 7. Billing & Payment Analysis

### 7.1 Monthly Charge

In [35]:
df["Monthly Charge Group"] = pd.cut(
    df["Monthly Charge"],
    bins=[0, 40, 60, 80, 100, float("inf")],
    labels=["< $40", "$40-60", "$60-80", "$80-100", "> $100"]
)

In [ ]:
df.groupby("Monthly Charge Group", observed=True)["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 7.2 Payment Method

In [ ]:
df.groupby("Payment Method")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 7.3 Paperless Billing

In [ ]:
df.groupby("Paperless Billing")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 7.4 Offer

In [ ]:
pd.crosstab(
    df["Offer"].fillna("Missing"),
    df["Customer Status"],
    normalize="index"
) * 100

## 8. Customer Engagement

### 8.1 Referred a Friend

In [ ]:
df. groupby("Referred a Friend")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

### 8.2 Number of Referrals

In [ ]:
df.groupby("Number of Referrals")["Churn Label"].apply(
    lambda x: (x == "Yes").mean() * 100
)

## 9. Churn Reasons

### 9.1 Churn Category

In [ ]:
df[df["Churn Label"] == "Yes"]["Churn Category"].value_counts()

### 9.2 Churn Reason

In [ ]:
df[df["Churn Label"] == "Yes"]["Churn Reason"].value_counts()

## 10. Key Findings

### 1. Overall Churn
- Overall churn rate was **26.54%** across 7,043 customers.

### 2. Contract Type
- Month-to-month customers had the highest churn rate at **45.84%**.
- Two-year contract customers had the lowest churn rate at **2.55%**.

### 3. Customer Tenure
- Customers with **0–12 months** of tenure had the highest churn rate at **47.44%**.
- Customers with **61+ months** of tenure had the lowest churn rate at **6.61%**.

### 4. Internet Service
- Customers with Internet Service had a churn rate of **31.83%**, compared with **7.40%** among customers without Internet Service.
- Fiber Optic customers showed the highest churn rate among Internet Types at **40.72%**.

### 5. Monthly Charges
- Churn rates were higher among customers with monthly charges of **$60–100**.
- The **$80–100** group had the highest churn rate at **37.02%**.

### 6. Churn Categories
- **Competitor** was the largest reported churn category, accounting for approximately **45%** of churned customers.

### 7. Churn Reasons
- Competitor-related reasons were among the most frequently reported churn reasons.
- Support and service-related reasons were also observed among churned customers.

> These findings describe patterns and associations observed in the dataset and do not imply causal relationships.